FaunaDB.Client/Types/RefV.cs

```diff
@@ -0,0 +1,159 @@
+    /// <para>
+    /// See <see href="https://fauna.com/documentation/queries#values-special_types">FaunaDB Special Types</see>.
+    /// </para>
+    /// </summary>
+    public class RefV : Value
+    {
+        public string Id { get; }
+        public RefV Class { get; }
+        public RefV Database { get; }
+
+        public RefV(string id, RefV @class = null, RefV database = null)
+        {
+            Id = id;
+            Class = @class;
+            Database = database;
+
+            if (@class != null && !@class.IsClassRef())
+                throw new ArgumentException($"Argument '{nameof(@class)}' is not a class reference", nameof(@class));
+
+            if (database != null && !database.IsDatabaseRef())
+                throw new ArgumentException($"Argument '{nameof(database)}' it not a database reference", nameof(database));
+        }
+
+        protected internal override void WriteJson(JsonWriter writer)
+        {
+            var props = UnescapedObject.With(
+                "id", Id,
+                "class", Class,
+                "database", Database
+            );
+
+            writer.WriteObject("@ref", props);
+        }
+
+        public override bool Equals(Expr v)
+        {
+            var other = v as RefV;
+
+            return other != null &&
+                Id == other.Id &&
+                Class == other.Class &&
+                Database == other.Database;
+        }
+
+        protected override int HashCode() =>
+            new { Id, Class, Database }.GetHashCode();
+
+        public override string ToString()
+        {
+            var cls = Class != null ? $", class = {Class}" : string.Empty;
+            var db = Database != null ? $", database = {Database}" : string.Empty;
+
+            return $"{GetType().Name}(id = \"{Id}\"{cls}{db})";
+        }
+    }
+
+    static class RefExtensions
+    {
+        static bool IsBuiltIn(this RefV _ref) =>
+            _ref.Class == null && _ref.Database == null;
+
+        public static bool IsClassRef(this RefV cls)
+        {
+            if (IsBuiltIn(cls))
+                return true;
+
+            return cls.Class != null && IsClassRef(cls.Class);
+        }
+
+        public static bool IsDatabaseRef(this RefV db)
+        {
+            if (db.Class == BuiltIn.DATABASES)
+                return true;
+
+            return db.Database != null && IsDatabaseRef(db.Database);
+        }
+    }
+
+    /// <summary>
+    /// A Database reference. Terse version for <c>new RefV(id, BuiltIn.DATABASES, database)</c>
+    /// </summary>
+    public class DatabaseV : RefV
+    {
+        public DatabaseV(string id, DatabaseV database = null) : base(id, BuiltIn.DATABASES, database)
+        { }
+    }
+
+    /// <summary>
+    /// A Class reference. Terse version for <c>new RefV(id, BuiltIn.CLASSES, database)</c>
+    /// </summary>
+    public class ClassV : RefV
+    {
+        internal ClassV(string builtin) : base(builtin, null, null)
```

Comment:
I would remove it or move it to `RefV`, if needed. Having it here is a little confusing, even as a internal constructor.